### PR TITLE
Sharding

### PR DESCRIPTION
**Pull request by minhtriet**

### CODE REVIEW

File: graddft_qnn/Box_Plot.py [hunk 1]
```diff
@@ -0,0 +1,35 @@
+import matplotlib.pyplot as plt
```
Comment (minhtriet) [inline graddft_qnn/Box_Plot.py:1]: What is it for

File: graddft_qnn/main.py [hunk 2]
```diff
@@ -1,3 +1,5 @@
+import os
```
Comment (minhtriet) [inline graddft_qnn/main.py:1]: consider loading the  batch_size dynamically, this assume batch_size has to be >= 10, which may not be the case

File: graddft_qnn/main.py [hunk 3]
```diff
@@ -150,8 +154,8 @@
     test_loss = np.sqrt(aggregated_cost / len(dataset["test"]))
     logging.info(f"Test loss {test_loss}")
 
-    checkpoint_path = pathlib.Path().resolve() / pathlib.Path(filename).stem
-    qnnf.save_checkpoints(parameters, tx, step=n_epochs, ckpt_dir=str(checkpoint_path))
+    #checkpoint_path = pathlib.Path().resolve() / pathlib.Path(filename).stem
```
[the remark this answers is not in the document]
Comment (minhtriet) [reply graddft_qnn/main.py:157]: Then could you remove the comments and leave these 2 lines as before

File: graddft_qnn/Box_Plot.py (hunk 1 above)
Comment (wonagram) [reply graddft_qnn/Box_Plot.py:1]: Yeah Okay I will push it

File: graddft_qnn/qnn_functional.py [hunk 4]
```diff
@@ -34,6 +35,49 @@ def _regularize_grid(self, grid: Grid, num_qubits, grid_data):
         ).astype(jnp.float32)
         return interpolated
 
+    def _regularize_grid2(self, grid, num_qubits, grid_data):
```
[the remark this answers is not in the document]
Comment (minhtriet) [reply graddft_qnn/qnn_functional.py:38]: then could I ask you two things
1. Implement a test function for the regularize grid (not just the `_regularize_grid()` or `_regularize_grid_2()`. The idea is if someone think of another way to regularize the grid, then  In my mind this function should look like this
```
def test_regularize_grid():
  # load a grid from a file, you can run a simulation and copy the data of the original grid

  # call a regularizer

  # Physics constraints 1: The sum of charges of the grid before and after must be the same

  # Physics constraints 2: ....
```
2. Could you rename `_regularize_grid_2` to something more meaningful. After 1 month I wouldn't know what `2` is. A docstring is very much welcome

File: tests/text_regularize_grid_jax.py [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+    interpolated_grid_weights = regularizer(grid, n_qubits, grid.weights)  # (m,m,m)
+    downsampled_grid_weights = interpolated_grid_weights.reshape(-1)
+
+    # apply normalization
+    normalized_grid_weights, normalized_charge_density = apply_normalization(
+        grid.weights, downsampled_grid_weights,
+        unscaled_coefficient_inputs, downsampled_charge_density
+    )
+
+    # invariants
```
[the remark this answers is not in the document]
Comment (wonagram) [reply tests/text_regularize_grid_jax.py:75]: The results(normalized physical values) are same, thus pass the test.

**Comment by wonagram** [comment]
The results are different, but not due to mistakes. Using PMAP parallelism simply splits the batch data across devices, and each device computes the loss. However, when computing losses in QNNFunctional, the code follows different paths because of the try/except logic. For example, the downsampling step uses different regularization functions. Even though I forced everything to use the second regularization function, there were still some differences. In theory, if we unify the path in QNNFunctional.py, the results should be the same.

**Comment by minhtriet** [comment]
> The results are different, but not due to mistakes. Using PMAP parallelism simply splits the batch data across devices, and each device computes the loss. However, when computing losses in QNNFunctional, the code follows different paths because of the try/except logic. For example, the downsampling step uses different regularization functions. Even though I forced everything to use the second regularization function, there were still some differences. In theory, if we unify the path in QNNFunctional.py, the results should be the same. And after computing the losses, they go through the same code that updates the model.

I don't think we can't raise the accuracy https://github.com/minhtriet/graddft-qnn/pull/44. Feel free to find the better hyperparameter
